alf/algorithms/sac_algorithm.py

```diff
@@ -543,45 +578,50 @@ def actor_loss_fn(dqda, action):
 
         actor_loss = nest.map_structure(actor_loss_fn, dqda, action)
         actor_loss = math_ops.add_n(nest.flatten(actor_loss))
-        alpha = torch.exp(self._log_alpha).detach()
         actor_info = LossInfo(
-            loss=actor_loss + alpha * log_pi,
-            extra=SacActorInfo(actor_loss=actor_loss, neg_entropy=log_pi))
+            loss=actor_loss + cont_alpha * continuous_log_pi,
+            extra=SacActorInfo(actor_loss=actor_loss, neg_entropy=neg_entropy))
         return critics_state, actor_info
 
     def _select_q_value(self, action, q_values):
         """Use ``action`` to index and select Q values.
         Args:
-            action: discrete actions with shape ``[batch_size]``.
-            critics: Q values with shape ``[batch_size, replicas, num_actions]``.
+            action (Tensor): discrete actions with shape ``[batch_size]``.
+            q_values (Tensor): Q values with shape ``[batch_size, replicas, num_actions]``.
         Returns:
-            selected Q values with shape ``[batch_size, replicas]``.
+            Tensor: selected Q values with shape ``[batch_size, replicas]``.
         """
         # action shape: [batch_size] -> [batch_size, n, 1]
         action = action.view(q_values.shape[0], 1, -1).expand(
             -1, q_values.shape[1], -1).long()
         return q_values.gather(-1, action).squeeze(-1)
 
     def _critic_train_step(self, exp: Experience, state: SacCriticState,
-                           action, log_pi):
+                           action, log_pi, action_distribution):
         critics, critics_state = self._compute_critics(
             self._critic_networks, exp.observation, exp.action, state.critics)
 
         target_critics, target_critics_state = self._compute_critics(
             self._target_critic_networks, exp.observation, action,
             state.target_critics)
+        target_critics = target_critics.min(dim=1)[0]
 
         if self._act_type == ActionType.Discrete:
             critics = self._select_q_value(exp.action, critics)
-            target_critics = self._select_q_value(action, target_critics)
+            target_critics = self._select_q_value(
+                action, target_critics.unsqueeze(dim=1))
+
         elif self._act_type == ActionType.Mixed:
             critics = self._select_q_value(exp.action[0], critics)
-            target_critics = self._select_q_value(action[0], target_critics)
-
-        target_critic = target_critics.min(dim=1)[0].reshape(log_pi.shape) - \
-                         (torch.exp(self._log_alpha) * log_pi)
+            discrete_act_dist = action_distribution[0]
+            target_critics = torch.sum(
+                discrete_act_dist.probs * target_critics, dim=-1)
 
-        target_critic = target_critic.detach()
+        target_critic = target_critics.reshape(exp.reward.shape)
+        entropy_reward = nest.map_structure(lambda la, lp: torch.exp(la) * lp,
```

Comment:
The name ``entropy_reward`` might be mis-leading. Might be better to move the negative sign from line 624 to here?